### PR TITLE
Fix for unexpected feature in hint handler where it doesn't allow empty hint for one side of T2T hint

### PR DESCRIPTION
**Pull request by ayobuenavista**

### CODE REVIEW

File: contracts/sol6/KyberHintHandler.sol [hunk 1]
```diff
@@ -207,6 +215,14 @@ abstract contract KyberHintHandler is IKyberHint, Utils5 {
         bytes32[] memory ethToTokenReserveIds,
         uint256[] memory ethToTokenSplits
     ) public view override returns (bytes memory hint) {
+        if (ethToTokenType == TradeType.BestOfAll) {
+            bytes32[] memory emptyReserveIds;
```
Comment (ilanDoron) [inline contracts/sol6/KyberHintHandler.sol:219]: can check the input values for Ids and splits are empty
and use them for building hint

File: contracts/sol6/KyberHintHandler.sol [hunk 2]
```diff
@@ -165,6 +165,14 @@ abstract contract KyberHintHandler is IKyberHint, Utils5 {
         bytes32[] memory tokenToEthReserveIds,
         uint256[] memory tokenToEthSplits
     ) public view override returns (bytes memory hint) {
+        if (tokenToEthType == TradeType.BestOfAll) {
```
Comment (ilanDoron) [inline contracts/sol6/KyberHintHandler.sol:168]: Check input splits and Ids are empty and use them for build

File: test/sol6/kyberNetwork.js [hunk 3]
```diff
@@ -71,8 +71,8 @@ let reserve;
 let numReserves;
 let info;
 let hint;
-const tradeTypesArray = [MASK_IN_HINTTYPE, MASK_OUT_HINTTYPE, SPLIT_HINTTYPE, EMPTY_HINTTYPE];
-const tradeStr = ["MASK IN", "MASK OUT", "SPLIT", "NO HINT"];
+const tradeTypesArray = [EMPTY_HINTTYPE, MASK_IN_HINTTYPE, MASK_OUT_HINTTYPE, SPLIT_HINTTYPE];
```
Comment (ilanDoron) [inline test/sol6/kyberNetwork.js:74]: can call it here best of all as well.

File: test/sol6/networkHelper.js [hunk 4]
```diff
@@ -33,10 +33,10 @@ const type_apr = "TYPE_APR";
 const type_MOCK = "TYPE_MOCK";
 const type_fpr = "TYPE_FPR";
 
-const MASK_IN_HINTTYPE = 0;
-const MASK_OUT_HINTTYPE = 1;
-const SPLIT_HINTTYPE = 2;
-const EMPTY_HINTTYPE = 3;
+const EMPTY_HINTTYPE = 0;
```
Comment (ilanDoron) [inline test/sol6/networkHelper.js:36]: change to bestOfall

File: contracts/sol6/IKyberHint.sol [hunk 5]
```diff
@@ -4,9 +4,10 @@ import "./IKyberReserve.sol";
 
 
 interface IKyberHint {
-    enum TradeType {MaskIn, MaskOut, Split}
+    enum TradeType {BestOfAll, MaskIn, MaskOut, Split}
     enum HintErrors {
         NoError, // Hint is valid
+        BestOfAllHintError, // reserveIDs and splits must be empty for BestOfAll hint
```
Comment (ilanDoron) [inline contracts/sol6/IKyberHint.sol:10]: seems like its inconsistent to add hint as part of the name.

its not done in any other error

Comment (ayobuenavista) [reply contracts/sol6/IKyberHint.sol]: Change to `NoneEmptyDataError`.

File: test/sol6/networkHelper.js (hunk 4 above)
Comment (ayobuenavista) [reply test/sol6/networkHelper.js]: Fixed

File: test/sol6/kyberNetwork.js (hunk 3 above)
Comment (ayobuenavista) [reply test/sol6/kyberNetwork.js]: Fixed

File: contracts/sol6/KyberHintHandler.sol (hunk 2 above)
Comment (ayobuenavista) [reply contracts/sol6/KyberHintHandler.sol]: Build is strict now, throwing error if reserveIds and splits is not empty.

File: contracts/sol6/IKyberHint.sol [hunk 6]
```diff
@@ -4,9 +4,10 @@ import "./IKyberReserve.sol";
 
 
 interface IKyberHint {
-    enum TradeType {MaskIn, MaskOut, Split}
+    enum TradeType {BestOfAll, MaskIn, MaskOut, Split}
     enum HintErrors {
         NoError, // Hint is valid
+        NoneEmptyDataError, // reserveIDs and splits must be empty for BestOfAll hint
```
Comment (Anyhowclick) [inline contracts/sol6/IKyberHint.sol:10]: Should it be NonEmptyDataError instead of NoneEmptyDataError

File: contracts/sol6/KyberHintHandler.sol [hunk 7]
```diff
@@ -165,6 +165,9 @@ abstract contract KyberHintHandler is IKyberHint, Utils5 {
         bytes32[] memory tokenToEthReserveIds,
         uint256[] memory tokenToEthSplits
     ) public view override returns (bytes memory hint) {
+        if (tokenToEthType == TradeType.BestOfAll)
+            checkReserveIdsSplitsIsEmpty(tokenToEthReserveIds, tokenToEthSplits);
```
Comment (ilanDoron) [inline contracts/sol6/KyberHintHandler.sol:169]: seems like this is checked in verifyData

do we need extra check here.
and in next function?

File: contracts/sol6/KyberHintHandler.sol [hunk 8]
```diff
@@ -311,6 +317,20 @@ abstract contract KyberHintHandler is IKyberHint, Utils5 {
         (t2eHint, e2tHint) = abi.decode(hint, (bytes, bytes));
     }
 
+    /// @notice Checks that the reserveId and splits arrays must be empty
+    /// @param reserveIds reserve IDs array
+    /// @param splits splits array
+    function checkReserveIdsSplitsIsEmpty(
+        bytes32[] memory reserveIds,
```
Comment (ilanDoron) [inline contracts/sol6/KyberHintHandler.sol:324]: seems function is over kill for two lines of code.

it makes 2 lines --> 10 lines

so the file as whole is harder to follow.

File: contracts/sol6/KyberHintHandler.sol (hunk 7 above)
Comment (ayobuenavista) [reply contracts/sol6/KyberHintHandler.sol]: Oh wow somehow we overlooked and I also forgot there's a verifyData function in build hint. Removed the duplicates.

File: contracts/sol6/KyberHintHandler.sol (hunk 8 above)
Comment (ayobuenavista) [reply contracts/sol6/KyberHintHandler.sol]: Yes, removed. Similar to earlier comment.

File: contracts/sol6/KyberHintHandler.sol [hunk 9]
```diff
@@ -406,6 +406,10 @@ abstract contract KyberHintHandler is IKyberHint, Utils5 {
         bytes32[] memory reserveIds,
         uint256[] memory splits
     ) internal pure returns (HintErrors) {
+        if (tradeType == TradeType.BestOfAll) {
+            if (reserveIds.length != 0 || splits.length != 0) return HintErrors.NonEmptyDataError;
```
Comment (Anyhowclick) [inline contracts/sol6/KyberHintHandler.sol:410]: `splits != 0` is checked in line 428, will throw the `HintErrors.SplitsNotEmptyError`.... So maybe we just need to check `reserveIds.length != 0`?

Comment (ilanDoron) [reply contracts/sol6/KyberHintHandler.sol:410]: lets continue with this version

we supply good enough error messages.
late for another change.

Comment (manhlx3006) [reply contracts/sol6/KyberHintHandler.sol:410]: I think this is more clear error type for BestOfAll case

File: test/sol6/kyberHintHandler.js [hunk 10]
```diff
@@ -140,6 +141,34 @@ contract('KyberHintHandler', function(accounts) {
...
+                expected = Helper.buildHint('BEST_OF_ALL')(t2eOpcode, [], []);
+        
+                Helper.assertEqual(hint, expected);
+            });
+
+            it('should revert the T2E BEST-OF-ALL HINT if reserveIds or splits is NOT EMPTY', async() => {
+                t2eOpcode = BEST_OF_ALL;
+                
+                await expectRevert(
+                    hintHandler.buildTokenToEthHint(
```
Comment (manhlx3006) [inline test/sol6/kyberHintHandler.js:162]: Do we have tests where (non-empty reserves + empty splits) or (empty reserves + non-empty splits)?

File: test/sol6/kyberNetwork.js [hunk 11]
```diff
@@ -27,7 +27,7 @@ const { expectEvent, expectRevert } = require('@openzeppelin/test-helpers');
 
 const {BPS, precisionUnits, ethDecimals, ethAddress, zeroAddress, emptyHint, zeroBN, MAX_QTY, MAX_RATE} = require("../helper.js");
 const {APR_ID, BRIDGE_ID, MOCK_ID, FPR_ID, type_apr, type_fpr, type_MOCK,
-    MASK_IN_HINTTYPE, MASK_OUT_HINTTYPE, SPLIT_HINTTYPE, EMPTY_HINTTYPE, ReserveType}  = require('./networkHelper.js');
+    MASK_IN_HINTTYPE, MASK_OUT_HINTTYPE, SPLIT_HINTTYPE, BEST_OF_ALL_HINTTYPE, ReserveType}  = require('./networkHelper.js');
```
Comment (manhlx3006) [inline test/sol6/kyberNetwork.js:30]: Should have tests with this type of hint error and see network throws expected revert

**Pull request merged**